### ISSUE TITLE
[SFT] What framework was used for supervised fine-tuning（SFT）?

### ISSUE THREAD

**Issue by chenqiuyuan** [apple/ml-diffucoder#9]

**Comment by summmeer** [comment]
Hi! 
We're using LLaMA-Factory to conduct SFT, with only minor changes (tailoring the loss computation to that of the diffusion model).

**Comment by chenqiuyuan** [comment]
> Hi! We're using LLaMA-Factory to conduct SFT, with only minor changes (tailoring the loss computation to that of the diffusion model).

Nice, will your SFT framework version be opensourced?

**Comment by summmeer** [comment]
Hi, I'd like to share the key implementation as follows.
```python
    def transition(self, x_0, sigma, maskable_mask):
        # move_chance = 1 - (-sigma).exp()
        move_chance = sigma
        move_indices = (torch.rand(*x_0.shape, device=x_0.device) < move_chance) & maskable_mask
        x_t = torch.where(move_indices, self.processing_class.mask_token_id, x_0)
        return x_t
        
    def diffusion_forward(self, model, x, src_mask, sampling_eps=1e-3):
        batch_size = x.shape[0]
        if src_mask is None:
            src_mask = torch.zeros_like(x, device=x.device, dtype=torch.bool)
        
        # Sample noise level
        t = (1 - sampling_eps) * torch.rand(x.shape[0], device=x.device) + sampling_eps
        
        # Compute sigma and dsigma
        sigma = t
        dsigma = torch.reciprocal(t)
        
        # Apply noise to input
        x_t = self.transition(x, sigma[:, None], maskable_mask=~src_mask)
        
        # Forward pass
        logits = model(input_ids=x_t, attention_mask=None).logits
        
        # Apply mask for loss computation
        loss_mask = x_t == self.processing_class.mask_token_id
        
        # Shift loss
        logits = logits[:, :-1]
        loss_mask = loss_mask[:, 1:]
        x_target = x[:, 1:]
        
        # Compute loss
        loss = F.cross_entropy(
            logits.reshape(-1, logits.shape[-1]), 
            x_target.reshape(-1), 
            reduction="none"
        ).float().reshape(batch_size, -1)
        
        loss = loss.masked_fill(~loss_mask, 0)
        final_loss = (dsigma[:, None] * loss).sum() / loss_mask.sum()
        unweighted_loss = loss.sum() / loss_mask.sum()
        
        return final_loss, unweighted_loss, x_t

    def _get_dynamic_ratio(self) -> float:
        """Calculate the dynamic ratio based on training progress.
        At the start (0% progress), ratio is near 0.
        At 2/3 of training, ratio becomes 1.
        """
        if not hasattr(self.state, 'global_step') or not hasattr(self.state, 'max_steps'):
            logger.warning("State not initialized, using default ratio")
            return 0.9  # fallback to default if state not initialized
        
        progress = self.state.global_step / self.state.max_steps
        if progress < 1/5:
            # Linear increase from 0 to 1 during first 1/5 of training
            return progress * 5  # multiply by 5 to reach 1 at 1/5 progress
        else:
            return 1.0

    @override
    def compute_loss(self, model, inputs, *args, **kwargs):
        # if torch.rand(1) < self._get_dynamic_ratio():
        #     src_mask = inputs["labels"] == IGNORE_INDEX
        # else:
        #     src_mask = None

        # # src_mask = None

        src_mask = inputs["labels"] == IGNORE_INDEX

        # Apply diffusion forward
        final_loss, unweighted_loss, x_t = self.diffusion_forward(
            model, 
            inputs["input_ids"], 
            src_mask
        )
        # wandb.log({"loss": final_loss.item(), "unweighted_loss": unweighted_loss.item()})
        
        return final_loss

    @override
    def prediction_step(
        self,
        model: "torch.nn.Module",
        inputs: dict[str, Union["torch.Tensor", Any]],
        prediction_loss_only: bool,
        ignore_keys: Optional[list[str]] = None,
        **gen_kwargs,
    ) -> tuple[Optional[float], Optional["torch.Tensor"], Optional["torch.Tensor"]]:
        r"""Remove the prompt part in the generated tokens.

        Subclass and override to inject custom behavior.
        """
        if self.args.predict_with_generate:  # do not pass labels to model when generate
            labels = inputs.pop("labels", None)
        else:
            labels = inputs.get("labels")
            
        # Create source mask where label is IGNORE_INDEX
        src_mask = labels == IGNORE_INDEX if labels is not None else None
        loss, _, x_t = self.diffusion_forward(model, inputs["input_ids"], src_mask)

        return loss, _, _
```

The above code snippet re-writes file `LLaMA-Factory/src/llamafactory/train/sft/trainer.py`. You might also need to modify the model/tokenizer loader in `LLaMA-Factory/src/llamafactory/model/loader.py`. 

```python
### in function def load_tokenizer(model_args: "ModelArguments") -> "TokenizerModule"
        tokenizer = AutoTokenizer.from_pretrained(
            model_args.model_name_or_path,
            **init_kwargs,
        )
        # add special tokens <|dlm_pad|>
        tokenizer.add_special_tokens({"pad_token": "<|dlm_pad|>"})
        print(f"Added special tokens: {tokenizer.pad_token}, {tokenizer.pad_token_id}")
### in def load_model
        load_class = AutoModel ## changed from AutoModelForCausalLM
```

**Comment by ethanbar11** [comment]
Hey! Great paper, and thank you so much for posting the training code patches for llama-factory.


I'm implementing the discrete diffusion training using the code you posted and am observing significantly higher loss values compared to standard Qwen fine-tuning. I'm training on code data (in instruction format) with sequences of max length 1024 (generating state sized 1024).

### Specific Impementation Questions:

1.
```python
# Current implementation
logits = model(input_ids=x_t, attention_mask=None).logits
loss_mask = (x_t == self.processing_class.mask_token_id)
```
Should padding tokens be explicitly excluded from both the diffusion masking process and loss computation? I noticed the original code passed `attention_mask=None` - is this intentional for discrete diffusion? And I haven't seen any concern regarding excluding padding from the loss computation/

2. Did you encounter high loss during training? I'm starting around ~50 when starting to train and it goes down to ~30, compared to Qwen2.5 Coder that starts around 0.5 and gets to 0.2...

Hopefully my questions makes sense. Would extremely appreciate any help.

**Comment by summmeer** [comment]
@ethanbar11 A common practice is regarding pad tokens as regular tokens during training, which is fine. You can also try to consider excluding padding tokens. The high loss at first is within expectation. Usually after 500-1000 steps, the loss will become very small (like 0.x).

**Comment by deadlykitten4** [comment]
Thanks for your wonderful work! 

What is the weight merge script you use? Is this also the official implementation provided by LLaMA-Factory?

I encounter a problem where the performance of dynamically loading the base model and lora weight during inference is normal, but once saved as a single model, the performance during inference is almost the same as without SFT. 

Hope you can help me with that :)

**Comment by summmeer** [comment]
@deadlykitten4 Hi. Yes I think we use the official code in LLaMA-Factory for lora ckpt loading (haven't try merging yet).
